### PR TITLE
feat: add nextjs example and supporting libraries

### PR DESCRIPTION
**Pull request by aeneasr**
Closes https://github.com/ory/integrations/issues/61

BREAKING CHANGES: This patch changes the options structure for `useSesssion`.

### CODE REVIEW

File: examples/nextjs-app-router/next/types.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+   */
+  override?: {
+    applicationName: string
+
+    /**
+     * Set this to use a custom login UI for the login flow. This path should be relative to the
+     * project root. Assuming you have a file at `./app/my-login/page.tsx`, you would set this to
+     * `/my-login`.
+     */
+    login_ui_path?: string
```
Comment (zepatrik) [inline examples/nextjs-app-router/next/types.ts:49]:
```suggestion
    loginUiPath?: string
```

File: examples/nextjs-app-router/next/types.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    /**
+     * Set this to use a custom registration UI for the registration flow. This path should be relative to the
+     * project root. Assuming you have a file at `./app/my-registration/page.tsx`, you would set this to
+     * `/my-registration`.
+     */
+    registration_ui_path?: string
```
Comment (zepatrik) [inline examples/nextjs-app-router/next/types.ts:56]:
```suggestion
    registrationUiPath?: string
```

File: examples/nextjs-app-router/next/types.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    /**
+     * Set this to use a custom recovery UI for the recovery flow. This path should be relative to the
+     * project root. Assuming you have a file at `./app/my-recovery/page.tsx`, you would set this to
+     * `/my-recovery`.
+     */
+    recovery_ui_path?: string
```
Comment (zepatrik) [inline examples/nextjs-app-router/next/types.ts:63]:
```suggestion
    recoveryUiPath?: string
```

File: examples/nextjs-app-router/next/types.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+    /**
+     * Set this to use a custom verification UI for the verification flow. This path should be relative to the
+     * project root. Assuming you have a file at `./app/my-verification/page.tsx`, you would set this to
+     * `/my-verification`.
+     */
+    verification_ui_path?: string
```
Comment (zepatrik) [inline examples/nextjs-app-router/next/types.ts:70]:
```suggestion
    verificationUiPath?: string
```

File: examples/nextjs-app-router/next/types.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    /**
+     * Set this to use a custom settings UI for the settings flow. This path should be relative to the
+     * project root. Assuming you have a file at `./app/my-settings/page.tsx`, you would set this to
+     * `/my-settings`.
+     */
+    settings_ui_path?: string
```
Comment (zepatrik) [inline examples/nextjs-app-router/next/types.ts:77]:
```suggestion
    settingsUiPath?: string
```

**Review by zepatrik — COMMENTED** [review]
Sorry, but this triggered me during the presentation 😉

**Comment by aeneasr** [comment]
🤣

**Comment by aeneasr** [comment]
@jonas-jonas why is it

```
import "@ory/elements-react/theme/styles.css"
```

and not

```
import "@ory/elements-react/theme/default/styles.css"
```

?

**Comment by aeneasr** [comment]
I'm stuck trying to create the `@ory/nextjs` library and importing them in the two examples. Kinda lost here, would appreciate some help @jonas-jonas

**Comment by jonas-jonas** [comment]
The stylesheet import was an oversight, fix is here:

@aeneasr https://github.com/ory/elements/pull/241

File: packages/nextjs/src/app/actions.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
+"use server"
+
+import { OryConfig } from "../"
+import { newOryProjectClient } from "../sdk"
+
+const client = newOryProjectClient()
+
+export async function getProjectConfig(): Promise<OryConfig> {
+  const list = await client.listProjects().catch((err) => {
```
[the remark this answers is not in the document]
Comment (aeneasr) [reply packages/nextjs/src/app/actions.ts]: I've removed it for now because it's currently unused. We can recover it from the git history.

File: packages/nextjs/src/pages/utils.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    searchParams.set(key, value)
+  })
+
+  return searchParams
+}
+
+export function useSearchParams() {
+  const router = useRouter()
+  return toSearchParams(router.query)
+}
```
[the remark this answers is not in the document]
Comment (aeneasr) [reply packages/nextjs/src/pages/utils.ts:55]: Only

```
// Copyright © 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0
"use client"
export { useRegistrationFlow } from "./registration"
export { oryPageRouterConfig } from "./config"
```

is exported in utils/index.js

File: packages/elements-react/src/theme/default/components/card/layout.tsx [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import { PropsWithChildren } from "react"
+
+export function DefaultCardLayout({ children }: PropsWithChildren) {
+  return (
+    <main className="p-4 pb-8 flex items-center justify-center flex-col gap-8 min-h-screen">
```
[the remark this answers is not in the document]
Comment (aeneasr) [reply packages/elements-react/src/theme/default/components/card/layout.tsx]: Agreed and this is how it's done now.

File: examples/nextjs-app-router/app/auth/layout.tsx [hunk 9]
```diff
@@ -0,0 +1,3 @@
+import { OryCardContent } from "@ory/elements-react"
```
[the remark this answers is not in the document]
Comment (aeneasr) [reply examples/nextjs-app-router/app/auth/layout.tsx]: @jonas-jonas it's now:

```
// app/auth/layout.tsx

// Copyright © 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { DefaultCardLayout } from "@ory/elements-react/theme"

export default DefaultCardLayout

```

**Comment by GWnbsp** [comment]
Am I being too impatient? The dependency @ory/nextjs for the example was not found, and indeed, it doesn't exist. Where should I search for it?

File: examples/nextjs-app-router/app/globals.css [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+/* SPDX-License-Identifier: Apache-2.0 */
+
+@tailwind base;
+@tailwind components;
+@tailwind utilities;
+
+body {
+  color: var(--foreground);
+  background: var(--background);
+  font-family: Arial, Helvetica, sans-serif;
```
Comment (jonas-jonas) [inline examples/nextjs-app-router/app/globals.css:11]: We use Inter in the designs, so the examples should, too.

File: examples/nextjs-pages-router/pages/api/hello.ts
Comment (jonas-jonas) [inline examples/nextjs-pages-router/pages/api/hello.ts:1]: what is this?

File: examples/nextjs-pages-router/pages/fonts/GeistMonoVF.woff
Comment (jonas-jonas) [inline examples/nextjs-pages-router/pages/fonts/GeistMonoVF.woff:1]: unused?

File: packages/elements-react/src/theme/default/components/card/layout.tsx
Comment (jonas-jonas) [inline packages/elements-react/src/theme/default/components/card/layout.tsx:1]: Didn't we want to not add this to ory/elements-react? But instead leave it to the consumer to define a layout?

File: packages/elements-react/src/theme/default/components/generic/page-header.tsx [hunk 11]
```diff
@@ -3,7 +3,7 @@
 
 import { OryPageHeaderProps, useComponents } from "@ory/elements-react"
 import { UserMenu } from "../ui/user-menu"
-import { useSession } from "@ory/elements-react/client"
+import { useSession } from "../../../../client"
```
Comment (jonas-jonas) [inline packages/elements-react/src/theme/default/components/generic/page-header.tsx:6]: revert, otherwise the theme bundle will contain a duplicate of `useSession` again, which leads to all sorts of issues.

File: packages/nextjs/README.md [hunk 12]
```diff
@@ -0,0 +1,17 @@
+# nextjs
```
Comment (jonas-jonas) [inline packages/nextjs/README.md:1]: Needs a readme.

File: packages/nextjs/src/utils/config.test.ts [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+  it("should use VERCEL_URL if __NEXT_PRIVATE_ORIGIN is not provided", () => {
+    ;(isProduction as jest.Mock).mockReturnValue(false)
+    process.env["VERCEL_URL"] = "vercel-url.com"
+    const config: Partial<OryConfig> = {}
+    const result = enhanceConfig(config)
+    expect(result.sdk.url).toBe("https://vercel-url.com")
+  })
+
+  xit("should use window.location.origin if VERCEL_URL is not provided", () => {
+    // Not sure if this works
```
Comment (jonas-jonas) [inline packages/nextjs/src/utils/config.test.ts:61]: What do we need to check to make sure this works?

File: packages/nextjs/src/utils/config.ts [hunk 14]
```diff
@@ -0,0 +1,48 @@
+// Copyright © 2024 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+import { OryConfig } from "../types"
+import { isProduction } from "./sdk"
+
+export function enhanceConfig(
```
Comment (jonas-jonas) [inline packages/nextjs/src/utils/config.ts:6]:
```suggestion
export function enhanceOryConfig(
```

And it needs docs.

File: packages/nextjs/src/utils/cookie.test.ts [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    )
+    expect(guessCookieDomain("https://localhost:1234/123", {})).toEqual(
+      "localhost",
+    )
+  })
+
+  test("understands public suffix list", () => {
+    expect(
+      guessCookieDomain(
+        "https://spark-public.s3.amazonaws.com/dataanalysis/loansData.csv",
```
Comment (jonas-jonas) [inline packages/nextjs/src/utils/cookie.test.ts:57]: Lol, where does this come from?

File: packages/nextjs/src/utils/rewrite.ts [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+        new URL(replaceWith, selfUrl).toString(),
+      )
+    }
+  }
+  return source.replaceAll(
+    matchBaseUrl.replace(/\/$/, ""),
+    new URL(selfUrl).toString().replace(/\/$/, ""),
+  )
+}
+export function rewriteJsonResponse<T extends object>(
```
Comment (jonas-jonas) [inline packages/nextjs/src/utils/rewrite.ts:35]: I think this needs tests. Seems like a pretty complicated function, maybe some docs/explanations would be good?

File: tsconfig.base.json [hunk 17]
```diff
@@ -9,6 +9,7 @@
     "skipLibCheck": true,
     "baseUrl": ".",
     "paths": {
+      "@ory/nextjs": ["packages/nextjs/src/index.ts"],
```
Comment (jonas-jonas) [inline tsconfig.base.json:12]: Don't think this is needed with npm workspaces, why did you add it?

File: packages/nextjs/src/utils/cookie.test.ts (hunk 15 above)
Comment (aeneasr) [reply packages/nextjs/src/utils/cookie.test.ts:57]: An LLM, maybe.

File: Makefile [hunk 18]
```diff
@@ -10,20 +10,24 @@ install:
 test:
 	npm run test
 
+build:
+	npm run build
+
+dev:
+	npm run build
+
```
[the remark this answers is not in the document]
Comment (aeneasr) [reply Makefile:18]: I've adjusted the scripts to now use the nx commands, which I always had to lookup in the GitHub ci config.

File: examples/nextjs-pages-router/pages/api/hello.ts
Comment (aeneasr) [reply examples/nextjs-pages-router/pages/api/hello.ts:1]: This is included in the default example. We should probably add an api route that verifies the cookie here?

File: packages/elements-react/src/theme/default/components/card/layout.tsx
Comment (aeneasr) [reply packages/elements-react/src/theme/default/components/card/layout.tsx:1]: Yeah, but at the same time not everyone has or knows tailwind, and I thought it's nice if we at least give a little helper to just get it working asap. I'm fine either way tbh, but I found this nice:

```
import { DefaultCardLayout } from "@ory/elements-react/theme"
import "@ory/elements-react/theme/styles.css"

export default DefaultCardLayout
```

File: packages/nextjs/src/utils/config.test.ts (hunk 13 above)
Comment (aeneasr) [reply packages/nextjs/src/utils/config.test.ts:61]: Somehow be able to modify window.location.origin which I have not been able to get working in this test

File: packages/nextjs/src/utils/rewrite.ts (hunk 16 above)
Comment (aeneasr) [reply packages/nextjs/src/utils/rewrite.ts:35]: We have several tests for this in `describe("rewriteJsonResponse", () => {`

I added docs

File: packages/nextjs/src/utils/config.ts (hunk 14 above)
Comment (aeneasr) [reply packages/nextjs/src/utils/config.ts:6]: Done!

File: tsconfig.base.json (hunk 17 above)
Comment (aeneasr) [reply tsconfig.base.json:12]: Done!
